protocols/relay/src/v2/copy_future.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Helper to interconnect two substreams, connecting the receiver side of A with the sender side of
+//! B and vice versa.
+//!
+//! Inspired by [`futures::io::Copy`].
```

Comment:
True. Though:

- `futures::io::Copy` only copies from `AsyncRead` `A` to `AsyncWrite` `B`, not vice-versa. We could use [`AsyncReadExt::split`](https://docs.rs/futures/0.3.18/futures/io/trait.AsyncReadExt.html#method.split) and then create two `futures::io::Copy`.

    (As an aside `AsyncReadExt::split` just uses a mutex under the hood. While I don't think the additional synchronization overhead would penalize us, I don't think it would be a performance optimization either.)

- `bytes_sent` is tracked for both directions, thus would require synchronization between the two shims.

None of the two are roadblocks, though, off the top of my head, I don't see a large opportunity for simplification. Not a strong opinion. Let me know in case you still want me to explore this option.

Reply:
Good point! It's a bit more involved than I imagined. so this is good as is :+1: